apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Converters/ObjectConverter.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright 2018, Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Cloud.Firestore.V1Beta1;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Cloud.Firestore.Converters
+{
+    /// <summary>
+    /// A converter for a target type of <see cref="Object"/>. This is a deserialize-only converter
+    /// (as there's nothing in Object to serialize to a value). It's effectively a "default deserializer",
+    /// picking the most appropriate type to deserialize to.
+    /// </summary>
+    internal sealed class ObjectConverter : IFirestoreInternalConverter
+    {
+        public object DeserializeMap(FirestoreDb db, IDictionary<string, Value> values) =>
+            values.ToDictionary(pair => pair.Key, pair => DeserializeValue(db, pair.Value));
+
+        public object DeserializeValue(FirestoreDb db, Value value)
+        {
+            switch (value.ValueTypeCase)
+            {
```

Comment:
Maybe use the switch only to "calculate" the target type and then do `ValueDeserializer.Deserialize(...)`?? It might be a little overhead to go through all of that, but the code in each of the cases is the same as the one in the corresponding converter, I think.

Reply:
Mmm, good point. (We could use an array lookup instead of the switch if we wanted, too.) Will merge with this as-is so we can get a clear comparison, but then see what it does to the benchmarks. I don't mind the duplication if it makes a significant difference, but I'd prefer to get rid of it, yes.